shotover/src/transforms/kafka/sink_cluster/scram_over_mtls.rs

```diff
@@ -12,11 +13,92 @@ use kafka_protocol::{
     messages::{ApiKey, CreateDelegationTokenRequest, RequestHeader},
     protocol::{Builder, StrBytes},
 };
-use rand::prelude::SliceRandom;
 use rand::rngs::SmallRng;
+use rand::{prelude::SliceRandom, SeedableRng};
 use serde::{Deserialize, Serialize};
-use std::{sync::Arc, time::Duration};
+use std::{collections::HashMap, sync::Arc, time::Duration};
 use tokio::sync::Notify;
+use tokio::sync::{mpsc, oneshot};
+
+pub struct TokenRequest {
+    username: String,
+    response_tx: oneshot::Sender<DelegationToken>,
+}
+
+#[derive(Clone)]
+pub struct TokenTask {
+    tx: mpsc::Sender<TokenRequest>,
+}
+
+impl TokenTask {
+    #[allow(clippy::new_without_default)]
+    pub fn new(
+        mtls_connection_factory: ConnectionFactory,
+        mtls_port_contact_points: Vec<String>,
+    ) -> TokenTask {
+        let (tx, mut rx) = mpsc::channel::<TokenRequest>(100);
+        tokio::spawn(async move {
+            loop {
+                match task(&mut rx, &mtls_connection_factory, &mtls_port_contact_points).await {
+                    Ok(()) => {
+                        // shotover is shutting down, terminate the task
+                        break;
+                    }
+                    Err(err) => {
+                        tracing::error!("Token task restarting due to failure, error was {err:?}");
+                    }
+                }
+            }
+        });
+        TokenTask { tx }
+    }
+
+    pub async fn get_token_for_user(&self, username: String) -> Result<DelegationToken> {
+        let (response_tx, response_rx) = oneshot::channel();
+        self.tx
+            .send(TokenRequest {
+                username,
+                response_tx,
+            })
+            .await?;
+        Ok(response_rx.await?)
+    }
+}
+
+async fn task(
+    rx: &mut mpsc::Receiver<TokenRequest>,
+    mtls_connection_factory: &ConnectionFactory,
+    mtls_port_contact_points: &[String],
+) -> Result<()> {
+    let mut rng = SmallRng::from_rng(rand::thread_rng())?;
+    let mut username_to_token = HashMap::new();
+    while let Some(request) = rx.recv().await {
+        let token = if let Some(token) = username_to_token.get(&request.username).cloned() {
+            token
+        } else {
+            let address =
+                KafkaAddress::from_str(mtls_port_contact_points.choose(&mut rng).unwrap())?;
+            let mut connection = mtls_connection_factory
+                // Must be unauthed since mTLS is its own auth.
+                .create_connection_unauthed(&address)
+                .await?;
+
+            let token =
+                create_delegation_token_for_user(&mut connection, request.username.clone()).await?;
+
+            // TODO: This sleep is currently load bearing...
+            //       Need to delay progression until token has propagated.
+            tokio::time::sleep(std::time::Duration::from_secs(4)).await;
```

Comment:
Don't need to address in this PR, but is there a better way to detect token propagation

Reply:
yep, figuring that out is my next priority